wandb/apis/workspaces/internal.py

```diff
@@ -0,0 +1,287 @@
+        updatedAt
+        createdAt
+        starCount
+        starred
+        parentId
+        locked
+        viewCount
+        showcasedAt
+        alertSubscription {
+            id
+            __typename
+        }
+        accessTokens {
+            id
+            token
+            view {
+            id
+            __typename
+            }
+            type
+            emails
+            createdBy {
+            id
+            username
+            email
+            name
+            __typename
+            }
+            createdAt
+            lastAccessedAt
+            revokedAt
+            projects {
+            id
+            name
+            entityName
+            __typename
+            }
+            __typename
+        }
+        __typename
+        }
+        """
+    )
+
+    api = wandb.Api()
+
+    if clone or not (name := view.name):
+        random_id = wandb.util.generate_id(11)
+        name = _workspaceify(random_id)
+
+    spec = view.viewspec.model_dump(by_alias=True, exclude_none=True)
+
+    # hack: We're re-using Report API objects.  In the Report API, the value here
+    # is expected to be None, but for Workspaces it's expected to be []
+    filters = spec["section"]["runSets"][0]["filters"]
+    if "filters" in filters and len(filters["filters"]) > 0:
+        filters["filters"][0]["filters"] = []
+
+    variables = {
+        "entityName": view.entity,
+        "projectName": view.project,
+        "name": name,
+        "displayName": view.display_name,
+        "type": "project-view",
+        "description": "",
+        "spec": json.dumps(spec),
+        "locked": False,
+    }
+
+    # Adding the view Id breaks stuff for reasons I don't understand
+    # if view.id:
+    #     variables["id"] = view.id
+
+    response = api.client.execute(query, variables)
+
+    return response
+
+
+def _get_view(entity: str, project: str, view_name: str) -> Dict[str, Any]:
+    # Use this query because it let you use view_name instead of id
+    query = gql(
+        """
+        query View($entityName: String, $name: String, $viewType: String = "runs", $userName: String, $viewName: String) {
+            project(name: $name, entityName: $entityName) {
+                allViews(viewType: $viewType, viewName: $viewName, userName: $userName) {
+                    edges {
+                        node {
+                            id
+                            displayName
+                            spec
+                        }
+                    }
+                }
+            }
+        }
+        """
+    )
+
+    api = wandb.Api()
+
+    response = api.client.execute(
+        query,
+        {
+            "viewType": "project-view",
+            "entityName": entity,
+            "projectName": project,
+            "name": project,
+            "viewName": _workspaceify(view_name),
+        },
+    )
+
+    edges = response.get("project", {}).get("allViews", {}).get("edges", [])
```

Comment:
I think the internal gql_client handles these for us, but yeah we should add Workspace-API-specific error messages

Reply:
I think the internals of the GQL client handle these errors.

Also, the server probably doesn't want to return 403 because it could be used to test existence of a project/view.